Support SQLAlchemy 2.0

phillipuniverse opened the pull request:
Fixes #83.

I am going to be running these in production soon and I figured I would go ahead and contribute back.

My process was to take the fork from @bodik who did most of the heavy lifting (thanks @bodik!) and then applied fixes to other SQLAlchemy 2.0 issues. My 1 change so far fixes this exception:

```
                    except (InvalidRequestError, AttributeError):
                    # query might not contain columns yet, hence cannot be compiled
                    # or query might be a sqla2.0 select statement
                    pass
                # also try to infer the models from various internals
    >           for table_tuple in query._setup_joins + query._legacy_setup_joins:
    E           AttributeError: 'Select' object has no attribute '_legacy_setup_joins'
```

Not exactly sure where to go from here, I haven't yet investigated the testing ecosystem of this library but probably we want to get SQLAlchemy 2.0 into the test suite.

@bodik @tomviner what do you think?

javimudi committed:
fix(models): Check for None entity when getting query models

Under some circunstances, such as prepping an aliased subquery to be joined later on, the model entity is missing,
provoking an error when trying to apply filters.

(cherry picked from commit edc7c244eeafc0e29722554f75050848608dad04)

bodik committed:
sqlalchemy 1.4 infer tables also from select object (fix auto_join)

(cherry picked from commit 203d4014c2f2937b4ffac321bad4cae1a48d15da)

bodik committed:
sqlalchemy 1.4 infer tables also from select object (sqlalchemy 2.0 style?)

(cherry picked from commit b6791e32af57ce2092c29d794546c89cbd52dfc8)

phillipuniverse committed:
Conditional access of _legacy_setup_joins property

This property is removed in SQLAlchemy 2.x

phillipuniverse commented:
Looking at the other open PRs, this also contains the fix to #67 by way of #63 (which @bodik had already applied to their fork).

phillipuniverse commented:
@tomviner FYI I rebased on the latest master to get the MariaDB pin to pass the GH workflows

tomviner commented:
> @tomviner FYI I rebased on the latest master to get the MariaDB pin to pass the GH workflows

Thanks for your PR, and rebase. The [tests for SQLA 1.4 are failing](https://github.com/juliotrigo/sqlalchemy-filters/actions/runs/7463205357?pr=84). Would you like to take a look at fixing them?

phillipuniverse commented:
@tomviner no problem I think I see the issue. Glad we’ve got good tests covering it!

I can also take a look at how to add SQLAlchemy 2 to CI

phillipuniverse committed:
Gitignore venv and local dev tool directories

phillipuniverse committed:
Tox with SQLAlchemy 2

phillipuniverse committed:
Add all compatible Python versions to .python-version and remove from .gitignore

This allows Tox to find all Python versions when using pyenv

phillipuniverse committed:
Move tox to default dev dependencies

phillipuniverse committed:
SQLAlchemy 2 compatibility with drop_all

drop_all() makes the 'bind' positional argument required

phillipuniverse committed:
Avoid building wheels on later Python versions with mysql and postgres drivers

phillipuniverse committed:
Configure Tox for SQLAlchemy 2 and Python 3.11, 3.12

Specify compatible versions of SQLAlchemy and SQLAlchemy-Utils for Python, SQLAlchemy versions

phillipuniverse committed:
SQLAlchemy 2 compatibility for hybrid properties and methods

phillipuniverse committed:
Move all Tox env configuration and version compatibility into tox.ini

phillipuniverse commented on .github/workflows/tests.yml:
```diff
@@ -30,41 +30,19 @@ jobs:
-      fail-fast: false
       matrix:
-        include:
-          # sqlalchemylatest (i.e. > 2.0.0) is not yet supported
-          # for any version of python
-
-          - {python: '3.7', tox: "py37-sqlalchemy1.0"}
-          - {python: '3.7', tox: "py37-sqlalchemy1.1"}
-          - {python: '3.7', tox: "py37-sqlalchemy1.2"}
-          - {python: '3.7', tox: "py37-sqlalchemy1.3"}
-          - {python: '3.7', tox: "py37-sqlalchemy1.4"}
-
-          - {python: '3.8', tox: "py38-sqlalchemy1.0"}
-          - {python: '3.8', tox: "py38-sqlalchemy1.1"}
-          - {python: '3.8', tox: "py38-sqlalchemy1.2"}
-          - {python: '3.8', tox: "py38-sqlalchemy1.3"}
-          - {python: '3.8', tox: "py38-sqlalchemy1.4"}
-
-          - {python: '3.9', tox: "py39-sqlalchemy1.0"}
-          - {python: '3.9', tox: "py39-sqlalchemy1.1"}
-          - {python: '3.9', tox: "py39-sqlalchemy1.2"}
-          - {python: '3.9', tox: "py39-sqlalchemy1.3"}
-          - {python: '3.9', tox: "py39-sqlalchemy1.4"}
-
-          # python3.10 with sqlalchemy <= 1.1 errors with:
-          # AttributeError: module 'collections' has no attribute 'MutableMapping'
-          - {python: '3.10', tox: "py310-sqlalchemy1.2"}
-          - {python: '3.10', tox: "py310-sqlalchemy1.3"}
-          - {python: '3.10', tox: "py310-sqlalchemy1.4"}
```
No longer necessary since this is now fully encoded in tox.ini.

phillipuniverse commented on .python-version:
```diff
@@ -0,0 +1,6 @@
+3.7.17
+3.8.18
+3.9.18
+3.10.13
+3.11.7
+3.12.1
```
I locally use pyenv and thought this would be useful to specify. Happy to remove if this muddies the waters for tooling.

phillipuniverse commented on setup.py:
```diff
@@ -31,9 +31,10 @@
+            'tox~=3.28'
         ],
-        'mysql': ['mysql-connector-python-rf==2.2.2'],
-        'postgresql': ['psycopg2==2.8.4'],
+        'mysql': ['mysql-connector-python-rf>=2.2.2'],
+        'postgresql': ['psycopg2>=2.8.4'],
```
I don't exactly recall the Python version that was building wheels. I didn't think the versions were all that important for validating proper execution. I can look into this deeper if necessary.

phillipuniverse commented on sqlalchemy_filters/models.py:
```diff
@@ -51,12 +56,15 @@ def _get_valid_field_names(self):
-    return orm_descriptor.extension_type == symbol('HYBRID_PROPERTY')
+    # SQLAlchemy 2 treats extension_type as an enum, not a symbol(). Enum is at sqlalchemy.ext.hybrid.HybridExtensionType
+
+    return str(orm_descriptor.extension_type) in ("symbol('HYBRID_PROPERTY')", 'HybridExtensionType.HYBRID_PROPERTY')
```
This is a little bit hacky but I was trying to optimize for performance. Other things I thought of:

- Use `isinstance()` to see if `orm_descriptor.extension_type` is an Enum and then use `.name` to compare. `isinstance()` can be really slow in a hot path, and this gets invoked a lot for all properties on queried models
- Try/except import since `HybridExtensionType` only exists in 2+. Kind of obtuse, I thought the same "hot path" reason to exclude this one too

Happy to change for a different opinion on the right way to check this.

phillipuniverse commented on test/conftest.py:
```diff
@@ -131,7 +131,7 @@ def connection(db_uri, db_engine_options, is_postgresql):
     yield connection
 
-    Base.metadata.drop_all()
+    Base.metadata.drop_all(engine)
```
SA 2.x compatibility

phillipuniverse commented on tox.ini:
```diff
@@ -10,11 +19,14 @@ extras =
     postgresql
 deps =
-    {py37,py38,py39,py310}: sqlalchemy-utils~=0.37.8
+    !sqlalchemy2,!sqlalchemylatest: sqlalchemy-utils~=0.37.8
```
I really wanted `!sqlalchemy{2,latest}` to work but it didn't. Alas.

phillipuniverse reviewed (commented):
@tomviner sorry it took me a bit but I have a fix up!

I also included a lot of changes to Tox and how tests run in CI. Probably the most impactful change is that with all the modifications I made to tox.ini, we can remove the Tox environment compatibility from the GitHub workflow config. The result is that we now have SQLAlchemy 2 running in CI and an easy to way to run it locally!

There are still a few test failures around loads and joins in SQLAlchemy 2. It will likely take me another week to get some bandwidth to fix these additional tests. Hopefully there's a straightforward fix!

phillipuniverse committed:
Fix tox expression for sqlalchemy-utils dep

phillipuniverse committed:
SQLAlchemy 2 swapped order of columns in SELECT statement

phillipuniverse committed:
Use tox version from setup.py

phillipuniverse committed:
Use Tox 3-compatible version of tox-gh-actions

phillipuniverse committed:
Fix flake8

phillipuniverse committed:
Add Python 3.11 and 3.12 to CI

phillipuniverse committed:
Minimum SQLAlchemy-Utils 0.38.3

Simplifies tox expressions and doesn't change anything fundamental about this library. 0.38.3 dropped support for Python 3.4 and 3.5 which sqlalchemy-filters never supported in the first place

trbtm commented:
Is there an ETA on this?

trbtm commented:
@tomviner is there any work open that I can support with in order to release this?

ConfusedCoder556 commented:
Any updates?

Ar0010r commented:
I there anything I could help with in order to merge this ?

yhbarzaga commented:
hi, thanks for the work on this!
do we have any ETA? there is anything we can do for help to release this update?

thanks in advance!

tomviner commented:
Hi, @juliotrigo and I tried to get the tests running on this branch. Couldn't see how to it on this PR.

So we branched off, pulled in master (which now also allows manual triggers of actions) and got this test result:
https://github.com/juliotrigo/sqlalchemy-filters/actions/runs/12953506107

The summary was 2 version of sqlalchemy succeeded, and 3 failed:
<img width="480" alt="image" src="https://github.com/user-attachments/assets/36bb45a6-4ddb-40c2-abaa-f36ac93dc252" />

henrikmathisen commented:
Happy birthday PR 84!

what's that? your _only_ wish is to be merged? well we have to ask mr @juliotrigo what we can do to make it happen

nimaxin commented:
Hey @juliotrigo, any chance we can get this PR reviewed and merged? It’s been open for a while, and a lot of folks are waiting on it.

juliotrigo commented:
Hi all, apologies for the delays. We noticed that tests were not automatically run in the pull request for each commit. @tomviner and I did some adjustments in `master` and tests should now by automatically running on the `master` branch.

However, as Tom mentioned: https://github.com/juliotrigo/sqlalchemy-filters/pull/84#issuecomment-2612939460, it seems that some tests are not passing on this branch. It'd be good if this remote branch could be updated with what's currently in `master`, which should trigger tests to be automatically run.

phillipuniverse commented:
@juliotrigo thanks for the update, I’ll get this branch updated soon!

phillipuniverse committed:
Merge branch 'master' into sqlalchemy-2.x

phillipuniverse commented:
Sorry everyone for all the back and forth on this, I re-opened under my personal fork as #91. I don't have a lot of updates just yet, I haven't had bandwidth to make much progress on the failing tests.

juliotrigo commented:
Thanks for the update @phillipuniverse You can re-open this pull request when you're ready and we'll take a look at it.
